### ISSUE TITLE
Buildship does not like config when using Gradle 4.1

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-lib#59]
```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':xtext-lib'.
> Could not resolve all files for configuration ':xtext-lib:classpath'.
   > Could not resolve org.eclipse.xtend:org.eclipse.xtend.lib:2.9.0.
     Required by:
         project :xtext-lib > org.xtext:xtext-gradle-plugin:1.0.18
      > Org.xtext:xtext-gradle-plugin:1.0.18 declares a dependency from configuration 'default' to configuration 'runtime' which is not declared in the descriptor for project :xtext-lib:org.eclipse.xtend.lib.
   > Could not resolve org.eclipse.xtend:org.eclipse.xtend.lib:2.10.0.
     Required by:
         project :xtext-lib > io.typefox.gradle:gradle-p2gen:0.1.0
      > Io.typefox.gradle:gradle-p2gen:0.1.0 declares a dependency from configuration 'default' to configuration 'runtime' which is not declared in the descriptor for project :xtext-lib:org.eclipse.xtend.lib.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
```

@oehme @spoenemann  do you have an idea what gradle/buildship trys to tell me?
for me it looks like bootstrapping is no longer working in buildship usecase

**Comment by cdietrich** [comment]
problem happens as well if i execute

`./gradlew build` in xtext-umbrella/releng/gradle-composite

**Comment by oehme** [comment]
Thanks for the detailed description. We did change quite a few things in composite builds, I'll have a look.

**Comment by cdietrich** [comment]
is it possible that the includeBuild command with a string and a file now behaves differently

**Comment by cdietrich** [comment]
it seems to be order dependent

we generate a locations.properties as follows

```
git.clone.xtext.lib.location=/Users/dietrich/tmp/bug/xtext-lib
git.clone.xtext.core.location=/Users/dietrich/tmp/bug/xtext-core
```

that determines the includeBuild order.

if you change the settings.gradle to keep that order without locations.properties as well, you can see it there too

```
if (!hasProperty('generatePrefs')) {
    def projects = getProjectsFromLocationProperties()
    if (!projects.isEmpty()) {
        projects.each {
            file ->
            logger.quiet("includeBuild ${file.toURI()}")
            includeBuild file
        }
    } else {
        // default behaviour if no locations.properties was generated by oomph
        (files { file('../../../').listFiles() })
        .filter { 
            File it ->
            it.isDirectory() && XTEXT_GRADLE_PROJECTS.split(',').any{p->it.name.contains(p)} && new File(it, "settings.gradle").exists()
        }.collect{ it.path }
        .toUnique { a, b ->
            findProjectName(a) <=> findProjectName(b)
        }.sort {
            XTEXT_GRADLE_PROJECTS.indexOf(findProjectName(it))
        }.each { it -> 
            logger.quiet("includeBuild ${it}/")
            includeBuild it+"/"
        }
    }     
}
```

**Comment by cdietrich** [comment]
if i revert the order i dont get an error

```
.sort {
            -1*XTEXT_GRADLE_PROJECTS.indexOf(findProjectName(it))
        }
```

**Comment by cdietrich** [comment]
bash script to reproduce

```
#!/bin/bash
git clone https://github.com/eclipse/xtext-umbrella.git
cd xtext-umbrella/
git checkout -b cd_gradle41 origin/cd_gradle41
cd ..
git clone https://github.com/eclipse/xtext-lib.git
cd xtext-lib/
git checkout -b cd_gradle41 origin/cd_gradle41
cd ..
git clone https://github.com/eclipse/xtext-core.git
cd xtext-core/
git checkout -b cd_gradle41 origin/cd_gradle41
cd ../
echo -e "git.clone.xtext.lib.location=$(pwd)/xtext-lib\ngit.clone.xtext.core.location=$(pwd)/xtext-core\n" > xtext-umbrella/locations.properties
cd xtext-umbrella/releng/gradle-composite/
./gradlew build
```

**Comment by oehme** [comment]
I checked and this is indeed caused by a change to how we discover plugin substitutions in composite builds. A project should never be substituted into its own `buildscript.classpath` though. I've opened a [bug report](https://github.com/gradle/gradle/issues/2723) for that and will make sure this gets prioritised.

**Comment by cdietrich** [comment]
same problem even with gradle 3.5 and

`./gradlew clean build --include-build ../xtext-lib`

executed in xtext-core

**Comment by cdietrich** [comment]
i did some experiments and did come up with

```
if (hasProperty('compositeBuild')) {
    includeBuild('../xtext-lib') {
        dependencySubstitution {
            all {
                if (it.requested instanceof ModuleComponentSelector) {
                    def mcs = it.requested as ModuleComponentSelector
                    if (mcs.group == 'org.eclipse.xtext' || mcs.group == 'org.eclipse.xtend') {
                        if (mcs.version == '2.14.0-SNAPSHOT') {
                            it.useTarget(project(":"+mcs.module))
                        }
                       
                    }
                }
            }
        }
    }
}
```

unfortunately i did not find a way to read `2.14.0-SNAPSHOT` from the property
@oehme do you have any idea? any hope to get the gradle bug for this fixed?

**Comment by cdietrich** [comment]
still an issue with gradle 4.7

**Comment by cdietrich** [comment]
other possible workaround:
get rid of dependency to xtext-gradle and p2 gen plugin in xbase lib

**Comment by cdietrich** [comment]
even more strage. seems to be dependendent on order of locations.properties
=> if xtext-core is imported before xtext-lib no problem :(

**Comment by cdietrich** [comment]
=> needs extensive testing with

```
    if (!projects.isEmpty()) {
        projects.reverse().each {
```
in umbrellas settings.gradle

**Comment by cdietrich** [comment]
@oehme do you have any idea why the order of "includeBuild" matters here?

**Comment by oehme** [comment]
If ordering makes a difference, that would be a bug. I'm getting more and more convinced we should make plugin dependency substitution explicit (e.g. by including a build in the `pluginManagement` block)  instead of the implicit default.

**Comment by cdietrich** [comment]
do you have a link on that?

**Comment by oehme** [comment]
https://github.com/gradle/gradle/issues/4343

**Comment by cdietrich** [comment]
@oehme this seems not to be working. 
if i add a pluginManagement to settings.gradle it still tries to load the flase plugins
mentioned from the dependencies section of the buildscript.
any further hints?

**Comment by cdietrich** [comment]
since there is no hope to get this fixed on gradle side or on xtext side (@oehme does not seem to have time for that) we will get rid of xtend and p2-gen plugins in xtext-lib and see if that helps)

**Comment by cdietrich** [comment]
as it looks the issue is solved with gradle 4.9+

**Comment by cdietrich** [comment]
confirmed fixed in gradle 4.9
